### PR TITLE
Add FieldsType field to ManagedFields to track content format

### PR DESCRIPTION
**Pull request by apelisse**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -1105,6 +1105,9 @@ type ManagedFieldsEntry struct {
 	// Time is timestamp of when these fields were set. It should always be empty if Operation is 'Apply'
 	// +optional
 	Time *Time `json:"time,omitempty" protobuf:"bytes,4,opt,name=time"`
+	// FieldsType identifies the format of raw bytes in Fields below.
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: "below" is terrible because it won't work in the docs :-)

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: names in the comment should match the json field names.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -1105,6 +1105,9 @@ type ManagedFieldsEntry struct {
...
+	// There is currently only one possible value: "json+v1"
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: I think you also need to document what happens if the object is json and the Fields are some binary format?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 3]
```diff
@@ -1105,6 +1105,9 @@ type ManagedFieldsEntry struct {
...
+	FieldsType string `json:"fieldsType,omitempty" protobuf:"bytes,6,name=fieldsType`
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Unions/one-of are a thing in our apis (this was my suggested versioning mechanism). RawExtensions where the contents are self-describing (e.g., via apiVersion+kind fields in JSON or via envelope in proto) are a thing (this was @deads2k's suggested mechanism).

However, having one field explain the type of another field is not a thing as far as I know? Can you point to an example of this? I think both of the other patterns are better.

**Comment by deads2k** [comment]
> Unions/one-of are a thing in our apis (this was my suggested versioning mechanism). RawExtensions where the contents are self-describing (e.g., via apiVersion+kind fields in JSON or via envelope in proto) are a thing (this was @deads2k's suggested mechanism).

I think the code has been updated with a union as the intent, but I don't see the discriminator field which will ensure that old clients can see they don't speak the level being used.

**Comment by apelisse** [comment]
> I think the code has been updated with a union as the intent, but I don't see the discriminator field which will ensure that old clients can see they don't speak the level being used.

`FieldsType` is meant to be the discriminator now

**Comment by apelisse** [comment]
> FieldsType is meant to be the discriminator now

https://github.com/kubernetes/kubernetes/pull/81754/commits/3e938ae626f5d68eadbb60ff932acd5c5f1f8932#diff-eca3b8d856fa2e661f6da91b61de5e76R1108

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 4]
```diff
@@ -1105,9 +1105,12 @@ type ManagedFieldsEntry struct {
...
 	// +optional
 	Time *Time `json:"time,omitempty" protobuf:"bytes,4,opt,name=time"`
-	// Fields identifies a set of fields.
+	// FieldsType is the discriminator for the different fields format and version.
+	// There is currently only one possible value: "FieldsJSONv1"
+	FieldsType string `json:"fieldsType,omitempty" protobuf:"bytes,6,opt,name=fieldsType"`
+	// FieldsJSONv1 holds the first JSON version format as described in the "Fields" type.
 	// +optional
-	Fields *Fields `json:"fields,omitempty" protobuf:"bytes,5,opt,name=fields,casttype=Fields"`
+	FieldsJSONv1 *FieldsJSONv1 `json:"fieldsJSONv1,omitempty" protobuf:"bytes,5,opt,name=fieldsJSONv1,casttype=FieldsJSONv1"`
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: You need to tombstone the original `fields` value and at first glance, the `FieldsJSONv1` doesn't look compatible with the previously shipped protobuf value from 1.15 here: https://github.com/kubernetes/kubernetes/blob/release-1.15/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1133.

@liggitt is there some proto-magic that makes delta from 1.15 compatible that I'm not seeing?

**Comment by deads2k** [comment]
> > I think the code has been updated with a union as the intent, but I don't see the discriminator field which will ensure that old clients can see they don't speak the level being used.
> 
> `FieldsType` is meant to be the discriminator now

apologies, didn't read far enough down.

**Comment by deads2k** [comment]
As long as we have proper compatibility, I'm ok with this.  I think you should consider tombstoning the old values for simplicity.

**Pull request merged**